### ISSUE TITLE
rpath has hard-coded path

### ISSUE THREAD

**Issue by paciorek** [nimble-dev/nimble#858]
We are now failing a CRAN install, because of a new check they are in the process of putting in place.
Here is info on the new check: https://developer.r-project.org/Blog/public/2019/02/14/staged-install/

Apparently they want packages to be installable such that they can be moved after installation. Since we hard code in the installation directory via -rpath (only on Linux not MacOS nor Windows), this is causing an issue.  The link above mentions we could try to opt out. However first we should:

1) discuss (remind) why we link libnimble.so into nimble.so and whether we need this
2) see if we can have rpath use a relative path

Here is the failure on Brian Ripley's machine:

```
* installing *source* package ‘nimble’ ...
** package ‘nimble’ successfully unpacked and MD5 sums checked
checking whether the C++ compiler works... yes
checking for C++ compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C++ compiler... yes
checking whether clang++ -std=gnu++11 accepts -g... yes
working on eigen
checking how to run the C++ preprocessor... clang++ -std=gnu++11 -E
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking Eigen/Dense usability... no
checking Eigen/Dense presence... no
checking for Eigen/Dense... no
Using the version of Eigen provided with NIMBLE
NIMBLE_DIR = /data/gannet/ripley/R/packages/tests-clang/nimble.Rcheck/00LOCK-nimble/00new/nimble
ENABLE_LIB=true
configure: creating ./config.status
config.status: creating inst/make/Makevars
config.status: creating inst/make/Makevars_lib
config.status: creating R/config.R
config.status: creating src/Makevars
removing debugging flags
make[1]: Entering directory '/data/gannet/ripley/R/packages/tests-clang/nimble/inst/CppCode'
/data/gannet/ripley/R/R-clang/bin/Rscript --vanilla ./customizeMakeconf.R
make[1]: Leaving directory '/data/gannet/ripley/R/packages/tests-clang/nimble/inst/CppCode'
Building libnimble
make[1]: Entering directory '/data/gannet/ripley/R/packages/tests-clang/nimble/inst/CppCode'
clang++ -std=gnu++11 -I"/data/gannet/ripley/R/R-clang/include" -DNDEBUG -DR_NO_REMAP -I"../include"  -isystem /usr/local/clang/include  -fpic  -O3 -Wall -pedantic -frtti  -c RcppUtils.cpp -o RcppUtils.o
clang++ -std=gnu++11 -I"/data/gannet/ripley/R/R-clang/include" -DNDEBUG -DR_NO_REMAP -I"../include"  -isystem /usr/local/clang/include  -fpic  -O3 -Wall -pedantic -frtti  -c EigenTypedefs.cpp -o EigenTypedefs.o
clang++ -std=gnu++11 -I"/data/gannet/ripley/R/R-clang/include" -DNDEBUG -DR_NO_REMAP -I"../include"  -isystem /usr/local/clang/include  -fpic  -O3 -Wall -pedantic -frtti  -c eigenUsingClasses.cpp -o eigenUsingClasses.o
clang++ -std=gnu++11 -I"/data/gannet/ripley/R/R-clang/include" -DNDEBUG -DR_NO_REMAP -I"../include"  -isystem /usr/local/clang/include  -fpic  -O3 -Wall -pedantic -frtti  -c RcppNimbleUtils.cpp -o RcppNimbleUtils.o
clang++ -std=gnu++11 -I"/data/gannet/ripley/R/R-clang/include" -DNDEBUG -DR_NO_REMAP -I"../include"  -isystem /usr/local/clang/include  -fpic  -O3 -Wall -pedantic -frtti  -c Utils.cpp -o Utils.o
clang++ -std=gnu++11 -I"/data/gannet/ripley/R/R-clang/include" -DNDEBUG -DR_NO_REMAP -I"../include"  -isystem /usr/local/clang/include  -fpic  -O3 -Wall -pedantic -frtti  -c NamedObjects.cpp -o NamedObjects.o
clang++ -std=gnu++11 -I"/data/gannet/ripley/R/R-clang/include" -DNDEBUG -DR_NO_REMAP -I"../include"  -isystem /usr/local/clang/include  -fpic  -O3 -Wall -pedantic -frtti  -c ModelClassUtils.cpp -o ModelClassUtils.o
clang++ -std=gnu++11 -I"/data/gannet/ripley/R/R-clang/include" -DNDEBUG -DR_NO_REMAP -I"../include"  -isystem /usr/local/clang/include  -fpic  -O3 -Wall -pedantic -frtti  -c accessorClasses.cpp -o accessorClasses.o
clang++ -std=gnu++11 -I"/data/gannet/ripley/R/R-clang/include" -DNDEBUG -DR_NO_REMAP -I"../include"  -isystem /usr/local/clang/include  -fpic  -O3 -Wall -pedantic -frtti  -c dists.cpp -o dists.o
clang++ -std=gnu++11 -I"/data/gannet/ripley/R/R-clang/include" -DNDEBUG -DR_NO_REMAP -I"../include"  -isystem /usr/local/clang/include  -fpic  -O3 -Wall -pedantic -frtti  -c nimDists.cpp -o nimDists.o
clang++ -std=gnu++11 -I"/data/gannet/ripley/R/R-clang/include" -DNDEBUG -DR_NO_REMAP -I"../include"  -isystem /usr/local/clang/include  -fpic  -O3 -Wall -pedantic -frtti  -c dllFinalizer.cpp -o dllFinalizer.o
clang++ -std=gnu++11 -I"/data/gannet/ripley/R/R-clang/include" -DNDEBUG -DR_NO_REMAP -I"../include"  -isystem /usr/local/clang/include  -fpic  -O3 -Wall -pedantic -frtti  -c nimbleGraph.cpp -o nimbleGraph.o
clang++ -std=gnu++11 -I"/data/gannet/ripley/R/R-clang/include" -DNDEBUG -DR_NO_REMAP -I"../include"  -isystem /usr/local/clang/include  -fpic  -O3 -Wall -pedantic -frtti  -c smartPtrs.cpp -o smartPtrs.o
clang++ -std=gnu++11 -I"/data/gannet/ripley/R/R-clang/include" -DNDEBUG -DR_NO_REMAP -I"../include"  -isystem /usr/local/clang/include  -fpic  -O3 -Wall -pedantic -frtti  -c predefinedNimbleLists.cpp -o predefinedNimbleLists.o
clang++ -std=gnu++11 -I"/data/gannet/ripley/R/R-clang/include" -DNDEBUG -DR_NO_REMAP -I"../include"  -isystem /usr/local/clang/include  -fpic  -O3 -Wall -pedantic -frtti  -c nimOptim.cpp -o nimOptim.o
linking libnimble.so
Linux   
clang++ -std=gnu++11 -shared -o libnimble.so  RcppUtils.o EigenTypedefs.o eigenUsingClasses.o RcppNimbleUtils.o Utils.o NamedObjects.o ModelClassUtils.o accessorClasses.o dists.o nimDists.o dllFinalizer.o nimbleGraph.o smartPtrs.o predefinedNimbleLists.o nimOptim.o 
rm *.o
make[1]: Leaving directory '/data/gannet/ripley/R/packages/tests-clang/nimble/inst/CppCode'
make[1]: Entering directory '/data/gannet/ripley/R/packages/tests-clang/nimble/inst/CppCode'
rm -f Makeconf
make[1]: Leaving directory '/data/gannet/ripley/R/packages/tests-clang/nimble/inst/CppCode'
** libs
make[1]: Entering directory '/data/gannet/ripley/R/packages/tests-clang/nimble/src'
clang++ -std=gnu++11 -I"/data/gannet/ripley/R/R-clang/include" -DNDEBUG -DR_NO_REMAP -I../inst/include    -isystem /usr/local/clang/include  -fpic  -g -O3 -Wall -pedantic -frtti  -c nimble.cpp -o nimble.o
clang++ -std=gnu++11 -shared -L/usr/local/clang/lib64 -L/usr/local/lib64 -o nimble.so nimble.o -L../inst/CppCode -lnimble -Wl,-rpath=/data/gannet/ripley/R/packages/tests-clang/nimble.Rcheck/00LOCK-nimble/00new/nimble/CppCode -L/data/gannet/ripley/R/R-clang/lib -lRlapack -L/data/gannet/ripley/R/R-clang/lib -lRblas -lgfortran -lm -lquadmath
make[1]: Leaving directory '/data/gannet/ripley/R/packages/tests-clang/nimble/src'
make[1]: Entering directory '/data/gannet/ripley/R/packages/tests-clang/nimble/src'
make[1]: Leaving directory '/data/gannet/ripley/R/packages/tests-clang/nimble/src'
installing to /data/gannet/ripley/R/packages/tests-clang/nimble.Rcheck/00LOCK-nimble/00new/nimble/libs
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded from temporary location
** checking absolute paths in shared objects and dynamic libraries
ERROR: absolute paths in ‘libs/nimble.so’ include the temporary installation directory: please report to the package maintainer and use ‘--no-staged-install’
* removing ‘/data/gannet/ripley/R/packages/tests-clang/nimble.Rcheck/nimble’
```

**Comment by paciorek** [comment]
Here's the email notification we just got on this.

EmiStatR ProjectTemplate RVowpalWabbit Rblpapi bayesdfa bigMap bmlm 
distrEllipse franc gWidgets2 ganalytics glmmfields nimble polidata 
prcbench race rfishbase spdynmod trialr uavRst walker zonator

Staged installation is to be a new feature of R 3.6.0: it is not yet the 
default in R-devel but it is scheduled to become so on Mar 1.  To turn 
it on before then, set environment variable R_INSTALL_STAGED=true : this 
has been done for the fedora-clang results on CRAN.

Tomas Kalibera has written a blog about this which can be seen at
https://developer.r-project.org/Blog/public/2019/02/14/staged-install/index.html 
.  For most of these packages he has some analysis of the issues at
https://github.com/kalibera/rstagedinst/tree/master/hardcoded
and is happy to help with interpretation (he is Cc:ed here).

Please adjust your package to work with staged installation, if possible 
before Mar 1.  A stopgap measure is to opt out by adding

StagedInstall: no

in the DESCRIPTION file, and the CRAN team will do that for any 
remaining packages on which others depend (currently EmiStatR 
ProjectTemplate Rblpapi gWidgets2 nimble rfishbase) when the default is 
switched.

**Comment by paciorek** [comment]
From discussion of Chris and Perry;

nimble.so links to libnimble.so so that users can call (from R) functions like nimble's distribution functions. We avoid copying CppCode/*cpp into src simply to reduce size of nimble.so.

given this the -rpath is needed because when symbols in nimble.so are called via .Call at R run time there is no other way for R to know where libnimble.so is. It's not something that would be in a user's LD_LIBRARY_PATH or ld.so.conf.

Presumably there is no way to make -rpath use relative paths because at run time in R, how would we build in finding CppCode relative to wherever nimble is installed.

Proposed solution: copy CppCode files in for libnimble.so as we do for libnimble.a in configure.
If we do this, we don't need PKG_LIBS put into src/Makevars and don't need rpath at all.

**Comment by paciorek** [comment]
We have a CRAN deadline of 3/15 to fix this issue - apparently simply opting out is not an option.

In branch remove-rpath, I modified configure.ac to copy from CppCode into src before compiling nimble.so and to remove the dynamic linking via rpath to libnimble.so.

This seems to work fine, but the size of nimble.so balloons from ~1 MB to ~20 MB, which is much larger than libnimble.so because of the -g flag. I doubt there is a way to address this. I'll note that while CRAN flags large package tarballs, that 20 MB is in the right tail of sizes of .so's for other R packages but there are a number of other packages that are that size or even larger, so I don't see a real cause for concern here.

@perrydv let me know if you've had any further thoughts.

**Comment by paciorek** [comment]
Fixed by PR #867.
